SPEC.md

```diff
@@ -85,20 +85,40 @@ The operations that CNI plugins must support are:
   - All parameters should be the same as those passed to the corresponding add operation.
   - A delete operation should release all resources held by the supplied containerid in the configured network.
 
-- Report version
+- `GET`: Get container network configuration
+  - This action should return the same result as an `ADD` action for the same inputs. The result should not change over the lifetime of the container.
+  - The plugin should return an error if any general internal state is unexpected. For example, if the plugin's data storage is missing or corrupt, or it cannot talk to a controller (if it has one) controller, it should return an error.
+  - The plugin should not return an error if its expected sandbox state (eg interfaces, IP addresses, routes, etc) is not found, as subsequent plugins in the chain may alter this state.
+  - The plugin will be passed a `prevResult` field in its configuration JSON on stdin which contains the `Result` object which the plugin (or the chain of which the plugin is part) returned to the runtime 
+  regenerate as much of its `Result` object as it can, which may be modified by plugins in the chain. For example, if the plugin-created interface exists and has the expected IP address, the plugin should add the interface and address to its `Result`. If the plugin-created interface does not have the expected IP address, the plugin should add the interface (but not the expected IP address) to its `Result`.
```

Comment:
Text is garbled between these two lines "returned to the runtime\nregenerate as much"; perhaps a mis-edit.

Having said that I still don't love this part (I know we've gone round and round on it).

Is this statement of what the prevResult contains valid from the perspective of the second(+) plugin in the chain?  Can we provide instructions for what a plugin should do that are independent of its chain position?

Is it fair to say that a chain of plugins should return the prevResult, or an error?  There's a "should" in the first line, but this leaves some wiggle room, so I think it's important to close down whether and how a runtime needs to deal with getting a different non-error response to the ADD or if it's prohibited.

If it's prohibited then it might still make sense to allow plugins to modify it, and just turn chain result != chain prevResult into an error in libcni - that would allow a plugin to say "this isn't the IP I gave it" by changing that field, and then a later plugin that modified the IP can change it back.  Essentially, we're using the result structure to pass per field unexpected value errors through a chain; and then testing for them at the end.  This is quite complex though, so if we do want this we need some better guidance.

If we want chains to sometimes return a different result that's not an error; then it becomes quite hard I think.  Plugins need to pass everything through by default, because they can't rely on a later plugin being able to regenerate any specific field.

I think it all works, but I do think we need text clarifications.  I'm happy to suggest some text if we can discuss what the meaning is.